File: frontend/services/storefront.ts
import { api, type ApiEnvelope } from '@/lib/api';
import { mediaUrl } from '@/lib/media-url';

export interface StorefrontGalleryItem {
  id: string;
  url: string;
  category: string;
  sort_order: number;
  is_featured: boolean;
  caption?: string | null;
}

export interface StorefrontHighlight {
  title: string;
  description?: string | null;
}

export interface StorefrontMachine {
  id: string;
  name: string;
  brand?: string | null;
  description?: string | null;
  image_url?: string | null;
}

export interface StorefrontTeamMember {
  id: string;
  name: string;
  role: string;
  description?: string | null;
  photo_url?: string | null;
}

export interface StorefrontCertification {
  id: string;
  title: string;
  issuer?: string | null;
  image_url?: string | null;
}

export interface StorefrontVideo {
  id: string;
  title: string;
  url: string;
  video_type: string;
}

export interface StorefrontData {
  laundry_id: string;
  template_id: string;
  is_published: boolean;
  logo_url: string | null;
  cover_url: string | null;
  brand_primary: string | null;
  brand_secondary: string | null;
  tagline: string | null;
  brand_story: string | null;
  years_in_business: number | null;
  owner_name: string | null;
  contact_phone: string | null;
  whatsapp_number?: string | null;
  show_call?: boolean;
  show_whatsapp?: boolean;
  show_callback?: boolean;
  approval_status?: string;
  working_hours: Record<string, string> | null;
  pickup_radius_km: string | null;
  delivery_radius_km: string | null;
  facilities: string[];
  highlights: StorefrontHighlight[];
  gallery: StorefrontGalleryItem[];
  machines: StorefrontMachine[];
  team: StorefrontTeamMember[];
  certifications: StorefrontCertification[];
  videos: StorefrontVideo[];
  completeness_score: number;
}

export interface StorefrontTemplate {
  id: string;
  name: string;
  description: string;
  brand_primary: string;
  brand_secondary: string;
  sample_facilities: string[];
  sample_highlights: StorefrontHighlight[];
}

export interface PublicStorefront {
  storefront: StorefrontData;
  laundry: {
    id: string;
    name: string;
    slug: string;
    city: string;
    address_line: string;
    description: string | null;
    avg_rating: string;
    review_count: number;
    is_verified: boolean;
    services: Array<{
      id: string;
      name: string;
      category: string;
      unit: string;
      price_inr: string;
      is_active: boolean;
    }>;
  };
  orders_completed: number;
  services: PublicStorefront['laundry']['services'];
}

export function resolveStorefrontImage(url: string | null | undefined): string {
  if (!url) return '';
  return mediaUrl(url);
}

export async function getPartnerStorefront(): Promise<StorefrontData> {
  const { data } = await api.get<ApiEnvelope<StorefrontData>>('/partner/storefront');
  return data.data;
}

export async function updatePartnerStorefront(
  patch: Partial<StorefrontData>,
): Promise<StorefrontData> {
  const { data } = await api.put<ApiEnvelope<StorefrontData>>('/partner/storefront', patch);
  return data.data;
}

export async function applyStorefrontTemplate(templateId: string): Promise<StorefrontData> {
  const { data } = await api.post<ApiEnvelope<StorefrontData>>(
    '/partner/storefront/apply-template',
    { template_id: templateId },
  );
  return data.data;
}

export async function listStorefrontTemplates(): Promise<StorefrontTemplate[]> {
  const { data } = await api.get<ApiEnvelope<StorefrontTemplate[]>>('/partner/storefront/templates');
  return data.data;
}

export async function getStorefrontOptions(): Promise<{
  facilities: string[];
  gallery_categories: string[];
}> {
  const { data } = await api.get<ApiEnvelope<{ facilities: string[]; gallery_categories: string[] }>>(
    '/partner/storefront/options',
  );
  return data.data;
}

export async function uploadStorefrontImage(file: File): Promise<string> {
  const form = new FormData();
  form.append('file', file);
  const { data } = await api.post<ApiEnvelope<{ url: string }>>('/partner/storefront/upload', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return data.data.url;
}

export async function getPublicStorefront(laundryId: string): Promise<PublicStorefront> {
  const { data } = await api.get<ApiEnvelope<PublicStorefront>>(`/laundries/${laundryId}/storefront`);
  return data.data;
}

File: backend/app/schemas/storefront.py
"""Storefront builder API schemas."""

from __future__ import annotations

from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.laundry import LaundryDetailResponse, LaundryServiceResponse


class StorefrontGalleryItem(BaseModel):
    id: str
    url: str
    category: str = "store"
    sort_order: int = 0
    is_featured: bool = False
    caption: str | None = None


class StorefrontHighlight(BaseModel):
    title: str = Field(max_length=120)
    description: str | None = Field(default=None, max_length=500)


class StorefrontMachine(BaseModel):
    id: str
    name: str = Field(max_length=120)
    brand: str | None = Field(default=None, max_length=80)
    description: str | None = Field(default=None, max_length=500)
    image_url: str | None = Field(default=None, max_length=2000)


class StorefrontTeamMember(BaseModel):
    id: str
    name: str = Field(max_length=120)
    role: str = Field(max_length=80)
    description: str | None = Field(default=None, max_length=500)
    photo_url: str | None = Field(default=None, max_length=2000)


class StorefrontCertification(BaseModel):
    id: str
    title: str = Field(max_length=120)
    issuer: str | None = Field(default=None, max_length=120)
    image_url: str | None = Field(default=None, max_length=2000)


class StorefrontVideo(BaseModel):
    id: str
    title: str = Field(max_length=120)
    url: str = Field(max_length=2000)
    video_type: str = Field(default="youtube", max_length=40)


class StorefrontTemplateInfo(BaseModel):
    id: str
    name: str
    description: str
    brand_primary: str
    brand_secondary: str
    sample_facilities: list[str]
    sample_highlights: list[StorefrontHighlight]


class StorefrontResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    laundry_id: UUID
    template_id: str
    is_published: bool
    logo_url: str | None
    cover_url: str | None
    brand_primary: str | None
    brand_secondary: str | None
    tagline: str | None
    brand_story: str | None
    years_in_business: int | None
    owner_name: str | None
    contact_phone: str | None
    whatsapp_number: str | None = None
    show_call: bool = True
    show_whatsapp: bool = True
    show_callback: bool = True
    approval_status: str = "approved"
    working_hours: dict[str, str] | None
    pickup_radius_km: Decimal | None
    delivery_radius_km: Decimal | None
    facilities: list[str]
    highlights: list[StorefrontHighlight]
    gallery: list[StorefrontGalleryItem]
    machines: list[StorefrontMachine]
    team: list[StorefrontTeamMember]
    certifications: list[StorefrontCertification]
    videos: list[StorefrontVideo]
    completeness_score: int


class StorefrontUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    template_id: str | None = Field(default=None, max_length=40)
    is_published: bool | None = None
    logo_url: str | None = Field(default=None, max_length=2000)
    cover_url: str | None = Field(default=None, max_length=2000)
    brand_primary: str | None = Field(default=None, max_length=7)
    brand_secondary: str | None = Field(default=None, max_length=7)
    tagline: str | None = Field(default=None, max_length=300)
    brand_story: str | None = None
    years_in_business: int | None = Field(default=None, ge=0, le=100)
    owner_name: str | None = Field(default=None, max_length=120)
    contact_phone: str | None = Field(default=None, max_length=20)
    whatsapp_number: str | None = Field(default=None, max_length=20)
    show_call: bool | None = None
    show_whatsapp: bool | None = None
    show_callback: bool | None = None
    working_hours: dict[str, str] | None = None
    pickup_radius_km: Decimal | None = Field(default=None, ge=0, le=50)
    delivery_radius_km: Decimal | None = Field(default=None, ge=0, le=50)
    facilities: list[str] | None = None
    highlights: list[StorefrontHighlight] | None = None
    gallery: list[StorefrontGalleryItem] | None = None
    machines: list[StorefrontMachine] | None = None
    team: list[StorefrontTeamMember] | None = None
    certifications: list[StorefrontCertification] | None = None
    videos: list[StorefrontVideo] | None = None


class ApplyTemplateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    template_id: str = Field(max_length=40)


class PublicStorefrontResponse(BaseModel):
    storefront: StorefrontResponse
    laundry: LaundryDetailResponse
    orders_completed: int
    services: list[LaundryServiceResponse]
